web/shared/src/typeahead.ts

```diff
@@ -144,23 +152,47 @@ export function triage<T>(
             beginswithCaseSensitive.push(obj);
         } else if (lowerItem.startsWith(lowerQuery)) {
             beginswithCaseInsensitive.push(obj);
+        } else if (
+            lowerItem.includes(" " + lowerQuery) ||
+            lowerItem.includes("_" + lowerQuery) ||
+            lowerItem.includes("-" + lowerQuery) ||
+            lowerItem.includes("/" + lowerQuery)
+        ) {
+            // space, hyphen, dash and slash characters are considered word boundaries
+            wordBoundaryMatch.push(obj);
         } else {
             noMatch.push(obj);
         }
     }
 
     if (sorting_comparator) {
-        const non_exact_sorted_matches = [
+        const beginning_matches_sorted = [
             ...beginswithCaseSensitive,
             ...beginswithCaseInsensitive,
         ].sort(sorting_comparator);
         return {
-            matches: [...exactMatch.sort(sorting_comparator), ...non_exact_sorted_matches],
+            matches: [
+                ...exactMatch.sort(sorting_comparator),
+                ...beginning_matches_sorted,
+                ...wordBoundaryMatch.sort(sorting_comparator),
+            ],
             rest: noMatch.sort(sorting_comparator),
         };
     }
+    if (for_recipients) {
+        return {
```

Comment:
Let's have a comment here explaining why this is special (I know we have it at the top too).

Though it might be a better API to just have a `request_raw_matches` parameter, which if true, means we just return the 5 categories to the caller and it can do with them as it sees fit? That would be more generic than `for_recipients`, and thus require less documentation. I don't love the name, feel free to replace it with something clearer to you.

Reply:
Allowing the return value to be the raw categories sounds great to me!

Since allowing another return type from the same function is fussy in typescript, especially with the multiple optional parameters here which make function overloading hard, I'll make a separate prep commit for this refactor (which I'll figure out tomorrow). Hope it will work out!